### ISSUE TITLE
Potential memory leak when hiding components early.

### ISSUE THREAD

**Issue by CamilleDrapier** [quasarframework/quasar#11804]
**Describe the bug**

When toggling the visibility of some quasar component's parent just after/while they are rendered, we see that the memory used by said component is not freed after un-render. This seems to be affecting at least q-menu and q-img and especially when they are rendered under "keep-alive", but might affect more components/scenarios.

**Codepen/jsFiddle/Codesandbox (required or your issue may be closed)**
Sorry, I'm not sure a live environment is very practical to measure the memory difference as outside factors could make the measurement difficult. I have set up this repository that should be able to show some of the problems, so I hope this would be acceptable:
<https://github.com/CamilleDrapier/quasar-test>

There are multiple branches:

- with "q-menu":
  - `keep-alive-menu` seems to have a memory leak problem
  - `v-show-menu`, is a variation of the previous one and does not show memory problem as far as I know
- with "q-img":
  - `keep-alive-img` seems to have a memory leak problem
  - `v-show-img`, is a variation of the previous one and also seems to have a memory leak problem
  - `q-img`, is a simpler variation of the previous one and also seems to have a memory leak problem. In this case, on Firefox, the memory problem seem to be harder to detect as sometimes a "Cycle Collection" would happen, but on Chrome we can observe a constant increase in the application footprint as we click on "toggle".

**To Reproduce**

1. Run `quasar build; quasar serve -p 8081 dist/spa/`
2. Access <http://localhost:8081> with your favorite browser (preferably in private windows mode, so that extensions do not interfere with measurements)
3. Open the "Developer tools" of your browser on the "memory tab"
4. Click on the "garbage collector button" (Chrome) or the 'GC' button on the `about:memory` page (Firefox)
5. Record the initial memory (snapshot)
6. Click on the "toggle" button 10 times
7. Click on the "garbage collector button" (Chrome) or the 'GC' button on the `about:memory` page (Firefox)
8. Record the memory again (snapshot)
9. (Edit) Repeat steps 6,7,8 again multiple times

**Expected behavior**
The memory used by the application does not increase consistently each cycle of toggling the content on and off.

**Screenshots**

For example on the `keep-alive-menu` here is a view of the memory increase for multiple batches of 10 toggles.

![Screenshot_20211228_192957](https://user-images.githubusercontent.com/13419648/147557593-92e76616-fa8a-404a-a0b7-370678fd312c.png)

**Platform (please complete the following information):**
Quasar Version: 2.4.2
@quasar/app Version: 3.2.6

Quasar mode:

- [X] SPA
- [ ] SSR
- [ ] PWA
- [ ] Electron
- [ ] Cordova
- [ ] Capacitor
- [ ] BEX
- [ ] Vite Plugin

Additional Info:

- [ ] Typescript
- [ ] SFC usage with `<script setup>`
- [ ] Vuex
- [ ] Prettier (eslint)
- [X] Standard (eslint)

Tested on:

- [X] SPA
- [ ] SSR
- [ ] PWA
- [ ] Electron
- [ ] Cordova
- [ ] Capacitor
- [ ] BEX

OS: Linux (20.04.2-Ubuntu SMP)
Node: v16.13.1
NPM: 8.1.2
Yarn:
Browsers: Firefox/Chrome
iOS:
Android:
Electron:

**Additional context**

While upgrading our application from Quasar1 to Quasar2, we noticed a sizable increase in the memory footprint of some of our pages, especially for one page that displays a list of components: for each component (that is quite complex/deep) we determine after it is rendered (in our application it is on some "intersection event", but is this app I have simplified this to "mounted" for a similar result memory-wise) if it should be kept visible or not, if not we store its size and set that size to a dummy container and "unshow" the content. This seemed to be working fine in quasar1, but is problematic in the newest version.

If this sort of thing is something we should avoid doing at all in Q2/vue3, is there some documentation that could help us know this? Or is there a better way to achieve this behavior?

Thanks a lot for the work on quasar in general and on the Quasar2 in particular!

**Comment by rstoenescu** [comment]
I believe this is a Vue 3 issue in regards to the keep alive component, because I can reproduce the issue without using any Quasar components.

Would you be so kind to report this to the Vue repository? In the reproduction repo for them, avoid using any Quasar components (so as to demonstrate the problem is indeed in Vue).

```vue
// OtherComponent.vue

<template>
  <div>
    <span>Menu</span>
  </div>
</template>

<script>
export default {
  name: 'OtherComponent'
}
</script>
```

```vue
// KeepAliveComponent.vue

<template>
  <div>
    <keep-alive>
      <OtherComponent v-if="visible" />
    </keep-alive>
  </div>
</template>

<script>
import OtherComponent from './OtherComponent.vue'

export default {
  name: 'KeepAliveComponent',
  components: {
    OtherComponent
  },
  data () {
    return {
      visible: true
    }
  },
  mounted () {
    this.visible = false
  }
}
</script>
```

```vue
// page

<template>
  <input type="button" @click="toggle" value="toggle" />
  <div v-if="visible">
    <span>Visible</span>
    <div v-for="n in 10" :key="n">
      <KeepAliveComponent />
    </div>
  </div>
</template>

<script>
import { defineComponent } from 'vue'

import KeepAliveComponent from '../components/KeepAliveComponent.vue'

export default defineComponent({
  name: 'App',
  components: {
    KeepAliveComponent
  },
  data () {
    return {
      visible: false
    }
  },
  methods: {
    toggle () {
      this.visible = !this.visible
    }
  }
})
</script>
```

**Comment by rstoenescu** [comment]
Please feel free to reopen if I missed something.

**Comment by CamilleDrapier** [comment]
Thanks a lot for looking into this!

I tried reproducing the problem with only simple HTML components, but couldn't see the same memory increase I could observe with q-menu or q-img.

For example, if I compare the memory snapshots with Chrome on the `keep-alive-span` (based on your suggestion) and `keep-alive-menu` branches of the repo I created for this issue, here is what I observe when measuring memory every 10 toggles + force garbage collection:

|keep-alive-span|keep-alive-menu|
|--|--|
|![Screenshot_20211230_075336](https://user-images.githubusercontent.com/13419648/147709214-2a64a951-cdf1-415f-a6c2-3b113fb2f241.png)|![Screenshot_20211230_075454](https://user-images.githubusercontent.com/13419648/147709231-7ed7448d-e877-4fcc-9025-72afa560f92d.png)|


If I check the `img` `q-img`, and the `q-img-visible` branches, I can see similar results:

|img|q-img|q-img-visible|
|--|--|--|
|![Screenshot_20211230_075336](https://user-images.githubusercontent.com/13419648/147709819-6e2594fd-4a12-4ccd-877f-b85042b7df3a.png)|![Screenshot_20211230_075454](https://user-images.githubusercontent.com/13419648/147709824-bcf17b9a-def5-4ae5-9051-82073d966aea.png)|![Screenshot_20211230_080945](https://user-images.githubusercontent.com/13419648/147709829-fa703730-8eca-4567-bd1f-f4a190ee3ab2.png)|

Please note that here, this potential problem with "q-img" is not related to "keep-alive", so it might impact more real-life use-cases.

I realize that my "Steps to reproduce the behavior" is quite incomplete and misguiding, I think it is fine to see an increase in the memory usage after the first use/display of the components (probably? because of some initialization/caching; I'm not completely sure...), what is troubling is that the increase continues happening as the components continue being used: repeating step 6,7 and 8 again and again.

Would you be able to have another look into this, please? Sorry for the trouble :bow: 

PS: I think I cannot reopen the issue myself, if you think this is worth exploring more, in quasar, could you re-open this?

**Comment by rstoenescu** [comment]
I understand your concerns, but like I said, I can reproduce it without using any Quasar components too...
Don't have anything actionable here unless someone has any ideas.

**Comment by CamilleDrapier** [comment]
Thank you, I tried running this on my Windows computer, just in case it was showing different results, but it seemed to behave the same way as I have described this so far.

Just in case I have not been explaining some part of the reproduction steps well, I made a small recording of what I did to show the problem when running the small project in both the "keep-alive-span" branch (no quasar components) and the "keep-alive-menu" branch

|keep-alive-span|keep-alive-menu|
|--|--|
|https://user-images.githubusercontent.com/13419648/147731179-631b2ac9-4127-4a1c-bc68-d0905ab600d2.mp4|https://user-images.githubusercontent.com/13419648/147731184-56d1f2c6-0c9c-49f8-bc92-1386bee3a105.mp4|






Let me know if there is anything else I can do to progress on this issue.

**Comment by rstoenescu** [comment]
Can this be tested again? (Upgrade q/app) Some memory leaks have been fixed in Vue 3.2.27 and 3.2.25.
https://github.com/vuejs/core/commit/2e3e183b4f19c9e25865e35438653cbc9bf01afc
https://github.com/vuejs/vue-next/commit/ed9eb62e5992bd575d999c4197330d8bad622cfb

Please reopen if it's still happening. It means the root cause is another mem leak in Vue.

**Comment by CamilleDrapier** [comment]
Thank you, I have tested this again and could at least reproduce the problem with q-img. I tried digging a bit deeper in that component and comment out some part of the code (in the `node_modules`'s quasar QImg.js file) that could be leading to the memory leak, after a while I could identify that the `loading: "lazy"` seems to be the trigger of the problem.

I pushed two new branches to the repository I created earlier: `q-img-eager` (that shows that the problem does not appear with the quasar component when using eager loading) and `img-lazy` (that shows that the problem is there when using stock img with lazy loading).

Considering this, I opened an issue on the vue-next repository about this problem with img.

I'm still not sure what is the problem with q-menu so I'll try to see if this can still be reproduced with the latest version of quasar/vue and update this thread with what I found.

**Comment by rstoenescu** [comment]
The "loading" prop is assigned as an attribute to the native rendered image. So if I understood correctly and there's a memory leak by using it then it means the browser itself is at fault because it does not releases the memory when the native img is removed from DOM. So not sure how any framework can help because it's not their fault (Quasar or Vue).

In any case, until this is fixed in browser(s), you can use `loading="eager"`.

**Comment by CamilleDrapier** [comment]
Thank you, indeed using `loading="eager"` could be a temporary solution for this problem for us, until the underlying issue may be fixed. I tried using `loading="lazy"` on my project that is still using quasar1 (our migration to v2 is more or less blocked by this memory issue), and couldn't see the same memory increase, I also tried with a stock `img(loading="lazy")`, so I'm not sure if this is a browser problem... I might also have made some measurement errors, but I did double check this... let's see if someone can have a look at the issue I opened for that specific problem on vue-next...

In parallel, I was still trying to find some other sources of memory leak in the specific keep-alive / visible:false situation I am in, and after tracking down to QdIalog's "use-timeout.js" (which uses "onBeforeUnmount"), I also noticed that some of my custom components that use a basic `beforeUnmount` are causing some big memory bloat for obscure reasons, so I also opened an issue for this on vue-next repo.

I think you were then initially right, and none of this seems to be specifically related with quasar, so my apologies for making this issue here and taking your time about this. I'll try to see if the issues I found so far can be fixed in vue.js and will re-check afterwards if there are any leftover stragglers.
